### PR TITLE
isabelle: Use prebuilt z3

### PR DESCRIPTION
**Pull request by jvanbruegge**
###### Description of changes

Required because old versions of z3 are being marked as insecure

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jvanbruegge** [comment]
I've also added a `checkPhase` that tests if the SMT solver is not broken

**Comment by jvanbruegge** [comment]
@shlevy could you have a look what is wrong with MacOS? If not I will just mark this package as linux-only again as I have no way to test it.

**Comment by shlevy** [comment]
@hrajchert Can you test on macOS?

**Comment by jvanbruegge** [comment]
I've marked it as linux only for now, so we can get this merged. Once someone with MacOS can debug this we can reenable it there. @SuperSandro2000 could you review this?

**Comment by jwiegley** [comment]
I can give this a try on both x86_64 and aarch64 macOS.

**Comment by jwiegley** [comment]
@jvanbruegge Both builds failed with:
```
source root is Isabelle2022
chmod: cannot access 'Isabelle2022': No such file or directory
```

**Comment by jvanbruegge** [comment]
Yes, I saw that in CI, but I dont know why that is. Without a Mac I cant replicate that

**Comment by jvanbruegge** [comment]
I suspect this is because it tries to `cd` into the source root before `postUnpack`? I changed that based on @SuperSandro2000 review during the Isabelle2022 update

**Comment by jvanbruegge** [comment]
I tried reverting that change now, let's see if that builds

**Comment by jvanbruegge** [comment]
Ok, that indeed worked.

@SuperSandro2000 this could be merged now

File: pkgs/applications/science/logic/isabelle/default.nix [hunk 1]
```diff
@@ -68,25 +66,22 @@ in stdenv.mkDerivation rec {
...
 
-  buildInputs = [ polyml z3 veriT vampire eprover-ho nettools ]
+  buildInputs = [ polyml veriT vampire eprover-ho nettools ]
     ++ lib.optionals (!stdenv.isDarwin) [ java ];
 
-  sourceRoot = dirname;
+  sourceRoot = "${dirname}${lib.optionalString stdenv.isDarwin ".app"}";
+
+  doCheck = true;
+  checkPhase = "bin/isabelle build -v HOL-SMT_Examples";
```
Comment (SuperSandro2000) [inline pkgs/applications/science/logic/isabelle/default.nix:75]: This should ideally be multiline

**Comment by jwiegley** [comment]
Still fails on `aarch64-darwin`:
```
       > /private/tmp/nix-build-isabelle-2022.drv-0/Isabelle2022/lib/scripts/isabelle-platform: line 26: sw_vers: command not found
       > *** I/O error: Cannot run program "/private/tmp/nix-build-isabelle-2022.drv-0/Isabelle2022/contrib/bash_process-1.3/platform_x86_64-darwin/bash_process": error=86, Bad CPU type in executable
```

**Pull request merged**